Add start charge session action for blue current integration.

NickKoepr opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds a start charge session action that starts a charging session with a given charging card. When no charging card is provided, a session without charging card will start.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/39146
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

NickKoepr committed:
Working on adding start session button.

Currently, charge cards are loaded and added to the configuration flow. Eventually a user can change his preferred charge card in this screen.

This screen can also be used to change the API token in a later stadium (if this implementation is needed).

NickKoepr committed:
A charge card can be selected and will be used to make the start session request.

NickKoepr committed:
Fixed a failing test.

This test failed because the get_charge_cards method was not included in the patch list.

NickKoepr committed:
Written test for blue current options flow.

NickKoepr committed:
When a charge card has a name, this name is also shown inside the dropdown.

* Changed tests to test a charge card with a name and without.
* BCU-APP charge card is now called 'None'.

NickKoepr committed:
Added translation key for selector schema for the options flow.

Now, 'Without charge card' is displayed for when a user does not want to use a charge card.

NickKoepr committed:
Added extra test for the charge cards and the add_charge_cards function is now also covered by the config flow test.

Using the init_integration function now, that generated a client mock that returns a mock charge card message back.

joostlek reviewed (changes requested):
Wouldn't it make more sense to have a service call for this instead? The target use case for this feature is for people to select one of their multiple cards, but the fact that they have multiple cards can also lead to the fact that they want to use both at the same time. By putting this in the options flow this is not automatable.

So I would either
1. Create 1 service and have the card as parameter
2. Create a separate button for every card

NickKoepr committed:
Added service for starting a charge session.

A user can also provide a charge card ID, or nothing to charge card defined in the config flow.

NickKoepr committed:
Changed charge card to charging card.

NickKoepr committed:
Moved the start charge session action from the button to init.

Now, a user can specify a device instead of a button. This fits the use case better, because the start button was not needed to start a charge session.

NickKoepr committed:
Changed config to integration settings.

NickKoepr commented:
Instead of requiring a start button from a charge point device, I have added a device_id option where users can select their charge point device. I think this is a better solution, because the start button was not needed to start a charge session.

NickKoepr committed:
Merge branch 'dev' into features/blue-current-start-button

NickKoepr committed:
Added missing suggest_object_id.

joostlek commented:
Why do we now implement both a service and a button?

Will there ever be more data related to the charge cards exposed to HA?

NickKoepr committed:
Removed not needed start charge session button.

NickKoepr committed:
Removed icon and string for start button.

NickKoepr commented:
> Why do we now implement both a service and a button?
> 
> Will there ever be more data related to the charge cards exposed to HA?

Oops, my bad. Forgot to remove the start button. Have removed it now!

We currently do not have plans to expose more charging card data to HA.

NickKoepr committed:
Merge branch 'dev' into features/blue-current-start-button

NickKoepr committed:
Updated button snapshots.

joostlek commented on homeassistant/components/blue_current/button.py:
Should we revert these changes?

joostlek commented on homeassistant/components/blue_current/__init__.py:
```diff
@@ -53,9 +69,33 @@ async def async_setup_entry(
+
+        await connector.client.start_session(evse_id, charging_card_id)
+
+    hass.services.async_register(DOMAIN, START_CHARGE_SESSION, start_charge_session)
```
We should register the services in `async_setup`

NickKoepr replied:
So in the `async_setup_entry` function, I have access to the Connector object (the Connector is set to `runtime_data`).  But when I register the service via `async_setup`, I do not have access to that Connector. Do you know a solution for this? Because I was not able to find one.

joostlek commented on homeassistant/components/blue_current/config_flow.py:
```diff
@@ -76,3 +83,69 @@ async def async_step_reauth(
+                return str(card[CONF_ID])
+            return f"{card[CONF_NAME]} ({card[CONF_ID]})"
+
+        def get_card_id(card_name: str) -> str:
+            """Get the card id based on the display name."""
+            split = card_name.rsplit("(")
+            if len(split) == 1:
+                if split[0] == WITHOUT_CHARGING_CARD:
+                    return BCU_APP
+                return split[0]
+            return split[-1].strip(")")
+
+        cards = [card_display_name(card) for card in connector.charge_cards.values()]
+
+        current_charge_card_id = card_display_name(connector.selected_charge_card)
+
+        card_schema = vol.Schema(
+            {
+                vol.Required(CARD, default=current_charge_card_id): SelectSelector(
+                    SelectSelectorConfig(
+                        options=cards,
+                        mode=SelectSelectorMode.DROPDOWN,
+                        translation_key="select_charging_card",
+                    )
+                )
+            }
+        )
+
+        if user_input is not None:
+            selected_card = list(
+                filter(
+                    lambda card: card[CONF_ID] == get_card_id(user_input[CARD]),
+                    connector.charge_cards.values(),
+                )
+            )[0]
+
+            user_input[CARD] = selected_card
+            return self.async_create_entry(title=CARD, data=user_input)
+
+        return self.async_show_form(step_id="init", data_schema=card_schema)
```
Do we still need this now?

NickKoepr committed:
Rolled back buttons and the config flow.

The config flow is not needed anymore. Now the service will defaults to using no charging card.

NickKoepr committed:
Updated description for charging_card_id.

NickKoepr committed:
Merge branch 'dev' into features/blue-current-start-button

NickKoepr committed:
Fixed linter issues and merge conflict error.

NoRi2909 commented on homeassistant/components/blue_current/strings.json:
```diff
@@ -136,5 +146,28 @@
+  "services": {
+    "start_charge_session": {
+      "name": "Start charge session",
+      "description": "Start a new charge session.",
```
Small nitpick: We generally use third person singular for action descriptions to ensure correct translations.
Also we should add a few words so it's not just paraphrasing the name by including some info about mandatory fields:

```suggestion
      "description": "Starts a new charge session on a specified charge point.",
```

NickKoepr committed:
Add service registration to async_setup and start_charge_session description change.

joostlek commented on homeassistant/components/blue_current/__init__.py:
```diff
@@ -14,33 +14,45 @@
 OBJECT = "object"
 VALUE_TYPES = [CHARGEPOINT_STATUS, CHARGEPOINT_SETTINGS]
 
+CONFIG_SCHEMA = cv.empty_config_schema(DOMAIN)
```
```suggestion
CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)
```

joostlek commented on homeassistant/components/blue_current/__init__.py:
```diff
@@ -67,6 +79,47 @@ async def async_setup_entry(
+
+        await connector.client.start_session(evse_id, charging_card_id)
+
+    hass.services.async_register(DOMAIN, START_CHARGE_SESSION, start_charge_session)
```
Make sure you supply a schema. This way you don't have to check if device_id None for example

joostlek commented on homeassistant/components/blue_current/__init__.py:
```diff
@@ -67,6 +79,47 @@ async def async_setup_entry(
+        # Get the device based on the given device ID.
+        device = dr.async_get(hass).devices.get(device_id)
+
+        if device is None:
+            raise ServiceValidationError
+
+        config_entry = hass.config_entries.async_get_entry(
+            list(device.config_entries)[0]
+        )
+
+        if config_entry is None or config_entry.state is not ConfigEntryState.LOADED:
+            raise ServiceValidationError
```
Let's also add some meaningful error to the ServiceValidationError

joostlek commented on homeassistant/components/blue_current/__init__.py:
```diff
@@ -67,6 +79,47 @@ async def async_setup_entry(
+
+        connector = config_entry.runtime_data
+
+        # Get the evse_id from the identifier of the device.
+        evse_id = list(device.identifiers)[0][1]
```
There is no guarantee that the first identifier is the blue_current one, so rather iterate over it)

joostlek commented on homeassistant/components/blue_current/const.py:
```diff
@@ -8,6 +8,12 @@
+BCU_APP = "BCU-APP"
+WITHOUT_CHARGING_CARD = "without_charging_card"
+CHARGING_CARD_ID = "charging_card_id"
+START_CHARGE_SESSION = "start_charge_session"
```
```suggestion
SERVICE_START_CHARGE_SESSION = "start_charge_session"
```
Maybe clearer to recognize

joostlek commented on tests/components/blue_current/test_init.py:
```diff
@@ -103,3 +107,85 @@ async def test_connect_request_limit_reached_error(
+
+    await hass.services.async_call(
+        DOMAIN,
+        "start_charge_session",
```
```suggestion
        START_CHARGE_SESSION,
```

joostlek commented on tests/components/blue_current/test_init.py:
```diff
@@ -103,3 +107,85 @@ async def test_connect_request_limit_reached_error(
+        DOMAIN,
+        "start_charge_session",
+        {
+            "device_id": list(dr.async_get(hass).devices)[0],
```
device registry can be imported as a test fixture

joostlek commented on tests/components/blue_current/test_init.py:
```diff
@@ -103,3 +107,85 @@ async def test_connect_request_limit_reached_error(
+        DOMAIN,
+        "start_charge_session",
+        {
+            "device_id": list(dr.async_get(hass).devices)[0],
+            "charging_card_id": "TEST_CARD",
```
Use constants for the keys

joostlek commented on tests/components/blue_current/test_init.py:
```diff
@@ -103,3 +107,85 @@ async def test_connect_request_limit_reached_error(
+    get_entry_mock = MagicMock()
+    get_entry_mock.state = ConfigEntryState.SETUP_IN_PROGRESS
+
+    with (
+        patch.object(
+            hass.config_entries, "async_get_entry", return_value=get_entry_mock
+        ),
+        pytest.raises(ServiceValidationError),
+    ):
```
I think it's practically impossible to have no config entry when you use the config entry id from a device, so I would just test that a mock config entry is not loaded

joostlek reviewed (dismissed):
Let's also update the title and description to match what the PR is doing

NickKoepr committed:
Applied received feedback.

emontnemery commented:
@NickKoepr What's this PR waiting for? It seems you've addressed the review comments?

emontnemery commented on homeassistant/components/blue_current/__init__.py:
```diff
@@ -67,6 +88,53 @@ async def async_setup_entry(
+                translation_domain=DOMAIN, translation_key="invalid_device_id"
+            )
+
+        config_entry = hass.config_entries.async_get_entry(
+            list(device.config_entries)[0]
+        )
```
This is not correct, because devices can (unfortunately) be connected to multiple config entries.
Instead, we need to find the `blue_current` config entry.
Maybe something like:
```py
blue_current_config_entry: ConfigEntry | None = None
for config_entry_id in device.config_entries:
    config_entry = hass.config_entries.async_get_entry(config_entry_id)
    if not config_entry or config_entry.domain != DOMAIN:
        # Not the blue_current config entry
        continue

    if config_entry.state is not ConfigEntryState.LOADED:
        raise ServiceValidationError(
            translation_domain=DOMAIN, translation_key="config_entry_not_loaded"
        )

     blue_current_config_entry = config_entry
     break

if not config_entry:
    # The device is not connected to a valid blue_current config entry
    raise ServiceValidationError(
        ...
    )
```

NickKoepr replied:
Fixed!

emontnemery reviewed (commented):
Looks good to me, just needs an improved way of finding the blue_current config entry

NickKoepr committed:
Merge branch 'dev' into features/blue-current-start-button

NickKoepr committed:
Added blue_current domain check when retrieving the config entry.

emontnemery reviewed (approved):
LGTM, thanks @NickKoepr 👍
